File: Chap1.R
# Chap1 R代码
# 自动从chap1.html同步生成

## 二项分布的概率分布列和分布函数
x <- seq(0, 9, 1)

# 概率分布列
plot(x, dbinom(x, size = 9, prob = 0.1), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "p(k)")
lines(x, dbinom(x, size = 9, prob = 0.2), col = 2, lty = 2)
lines(x, dbinom(x, size = 9, prob = 0.3), col = 3, lty = 3)
lines(x, dbinom(x, size = 9, prob = 0.5), col = 4, lty = 4)

# 累积分布函数
plot(x, pbinom(x, size = 9, prob = 0.1), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "F(k)")
lines(x, pbinom(x, size = 9, prob = 0.2), col = 2, lty = 2)
lines(x, pbinom(x, size = 9, prob = 0.3), col = 3, lty = 3)
lines(x, pbinom(x, size = 9, prob = 0.5), col = 4, lty = 4)

## 泊松分布的概率分布列和分布函数
x <- seq(0, 10, 1)

# 概率分布列
plot(x, dpois(x, lambda = 1), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "p(k)")
lines(x, dpois(x, lambda = 2), col = 2, lty = 2)
lines(x, dpois(x, lambda = 3), col = 3, lty = 3)
lines(x, dpois(x, lambda = 5), col = 4, lty = 4)

# 累积分布函数
plot(x, ppois(x, lambda = 1), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "F(k)")
lines(x, ppois(x, lambda = 2), col = 2, lty = 2)
lines(x, ppois(x, lambda = 3), col = 3, lty = 3)
lines(x, ppois(x, lambda = 5), col = 4, lty = 4)

## 负二项分布：size = r, prob = 1/(1+beta)
x <- seq(0, 10, 1)

# 概率分布列
plot(x, dnbinom(x, size = 2, prob = 1/(1+0.1)), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "p(k)")
lines(x, dnbinom(x, size = 2, prob = 1/(1+0.2)), col = 2, lty = 2)
lines(x, dnbinom(x, size = 2, prob = 1/(1+0.3)), col = 3, lty = 3)
lines(x, dnbinom(x, size = 2, prob = 1/(1+0.5)), col = 4, lty = 4)

# 累积分布函数
plot(x, pnbinom(x, size = 2, prob = 1/(1+0.1)), type = "l", col = 1, lty = 1,
     xlab = "k", ylab = "F(k)")
lines(x, pnbinom(x, size = 2, prob = 1/(1+0.2)), col = 2, lty = 2)
lines(x, pnbinom(x, size = 2, prob = 1/(1+0.3)), col = 3, lty = 3)
lines(x, pnbinom(x, size = 2, prob = 1/(1+0.5)), col = 4, lty = 4)

## 零截断负二项分布的概率
x <- 0:10
p <- dnbinom(x, 4, 0.7)       # 负二项分布的概率
p0 <- p[1]                    # 零点的概率
pt1 <- p[2:11] / (1 - p0)     # 零截断分布在非0点上的概率
pt <- c(0, pt1)               # 零截断分布的概率
com <- rbind(负二项 = p, 零截断负二项 = pt)
barplot(com, beside = TRUE, names.arg = 0:10, legend.text = TRUE)

## 零调整负二项分布的概率
pm0 <- 0.3                    # 调整后的零点概率
pm1 <- (1 - pm0) * p[2:11] / (1 - p0)
pm <- c(pm0, pm1)
com <- rbind(负二项 = p, 零调整负二项 = pm)
barplot(com, beside = TRUE, names.arg = 0:10, legend.text = TRUE)

## 指数分布的概率密度函数
x <- seq(0, 5, 0.01)
plot(x, dexp(x, rate = 0.5), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dexp(x, rate = 1), col = 2, lty = 2)
lines(x, dexp(x, rate = 2), col = 3, lty = 3)
lines(x, dexp(x, rate = 5), col = 4, lty = 4)

## 伽马分布概率密度函数：改变形状参数
x <- seq(0, 4, 0.001)
plot(x, dgamma(x, shape = 1, scale = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dgamma(x, shape = 2, scale = 1), col = 2, lty = 2)
lines(x, dgamma(x, shape = 3, scale = 1), col = 3, lty = 3)
lines(x, dgamma(x, shape = 0.5, scale = 1), col = 4, lty = 4)

## 伽马分布概率密度函数：改变尺度参数
plot(x, dgamma(x, shape = 1, scale = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dgamma(x, shape = 1, scale = 2), col = 2, lty = 2)
lines(x, dgamma(x, shape = 1, scale = 3), col = 3, lty = 3)
lines(x, dgamma(x, shape = 1, scale = 0.5), col = 4, lty = 4)

## 逆高斯分布概率密度函数（需 statmod 包）
x <- seq(0, 5, 0.01)

## 改变均值参数
plot(x, dinvgauss(x, mean = 1, shape = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dinvgauss(x, mean = 2, shape = 1), col = 2, lty = 2)
lines(x, dinvgauss(x, mean = 5, shape = 1), col = 3, lty = 3)
lines(x, dinvgauss(x, mean = 0.5, shape = 1), col = 4, lty = 4)

## 改变形状参数
plot(x, dinvgauss(x, mean = 1, shape = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dinvgauss(x, mean = 1, shape = 2), col = 2, lty = 2)
lines(x, dinvgauss(x, mean = 1, shape = 5), col = 3, lty = 3)
lines(x, dinvgauss(x, mean = 1, shape = 0.5), col = 4, lty = 4)

## 定义帕累托分布的概率密度函数
pareto.pdf <- function(x, alpha, lambda){
  alpha * lambda^alpha / (lambda + x)^(alpha + 1)
}

## 帕累托分布概率密度函数：改变形状参数
x <- seq(0, 3, 0.01)
plot(x, pareto.pdf(x, alpha = 1, lambda = 3), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, pareto.pdf(x, alpha = 2, lambda = 3), col = 2, lty = 2)
lines(x, pareto.pdf(x, alpha = 5, lambda = 3), col = 3, lty = 3)
lines(x, pareto.pdf(x, alpha = 0.5, lambda = 3), col = 4, lty = 4)

## 帕累托分布概率密度函数：改变尺度参数
plot(x, pareto.pdf(x, alpha = 2, lambda = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, pareto.pdf(x, alpha = 2, lambda = 2), col = 2, lty = 2)
lines(x, pareto.pdf(x, alpha = 2, lambda = 5), col = 3, lty = 3)
lines(x, pareto.pdf(x, alpha = 2, lambda = 0.5), col = 4, lty = 4)

## 对数正态分布概率密度函数：改变均值参数
x <- seq(0, 7, 0.01)
plot(x, dlnorm(x, meanlog = 1, sdlog = 0.5), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dlnorm(x, meanlog = 1, sdlog = 3), col = 2, lty = 2)
lines(x, dlnorm(x, meanlog = 1, sdlog = 10), col = 3, lty = 3)
lines(x, dlnorm(x, meanlog = 1, sdlog = 1), col = 4, lty = 4)

## 对数正态分布概率密度函数：改变方差参数
plot(x, dlnorm(x, meanlog = 1, sdlog = 0.5), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dlnorm(x, meanlog = 1, sdlog = 3), col = 2, lty = 2)
lines(x, dlnorm(x, meanlog = 1, sdlog = 10), col = 3, lty = 3)
lines(x, dlnorm(x, meanlog = 1, sdlog = 1), col = 4, lty = 4)

## 威布尔分布概率密度函数：改变形状参数
x <- seq(0, 3, 0.01)
plot(x, dweibull(x, shape = 1, scale = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dweibull(x, shape = 2, scale = 1), col = 2, lty = 2)
lines(x, dweibull(x, shape = 3, scale = 1), col = 3, lty = 3)
lines(x, dweibull(x, shape = 0.5, scale = 1), col = 4, lty = 4)

## 威布尔分布概率密度函数：改变尺度参数
plot(x, dweibull(x, shape = 1, scale = 1), type = "l", col = 1, lty = 1,
     xlab = "x", ylab = "f(x)")
lines(x, dweibull(x, shape = 1, scale = 2), col = 2, lty = 2)
lines(x, dweibull(x, shape = 1, scale = 3), col = 3, lty = 3)
lines(x, dweibull(x, shape = 1, scale = 0.5), col = 4, lty = 4)

## 定义混合分布的概率密度函数
f <- function(x){
  0.3 * dlnorm(x, meanlog = 1, sdlog = 2) + 0.7 * dlnorm(x, meanlog = 3, sdlog = 4)
}

curve(f, xlim = c(0, 1), ylim = c(0, 2), lty = 1, lwd = 2, col = 2)
curve(dlnorm(x, meanlog = 1, sdlog = 2), lty = 2, add = TRUE)
curve(dlnorm(x, meanlog = 3, sdlog = 4), lty = 3, add = TRUE)

set.seed(123)  # 设置随机种子以保证结果可重复
x <- rgamma(100, 2, 2.5)  # 从Ga(2, 2.5)中抽取100个随机数

## 极大似然法估计参数
fit1 <- fitdist(x, 'gamma', method = 'mle')
fit1  # 形状参数估计值约为2.19，速率参数估计值约为3.18
plot(fit1)

## 矩估计法估计参数
fit2 <- fitdist(x, 'gamma', method = 'mme')

## 分位数法估计参数
fit3 <- fitdist(x, 'gamma', method = 'qme', probs = c(1/3, 2/3))

## 最小距离法估计参数
fit4 <- fitdist(x, 'gamma', method = 'mge', gof = 'CvM')

summary(fit1)

## 卡方拟合优度检验：维修费用数据
lower <- c(0, 1000, 2000, 3000, 4000, 5000)
upper <- c(1000, 2000, 3000, 4000, 5000, Inf)
freq  <- c(250, 300, 250, 150, 100, 0)

## 指数分布的极大似然估计
f1 <- function(a) {-sum(freq * log(pexp(upper, a) - pexp(lower, a)))}
m1 <- optimize(f1, c(1e-5, 1e-3))

## 计算理论分布
q <- pexp(upper, m1$minimum) - pexp(lower, m1$minimum)
n <- sum(freq) * q

## 作卡方检验
chisq.test(freq, p = q)

## 泊松分布拟合检验
X <- 0:6
Y <- c(7, 10, 12, 8, 3, 2, 0)
q <- ppois(X, mean(rep(X, Y)))
n <- length(Y)
p <- rep(0, n)
p[1] <- q[1]
p[n] <- 1 - q[n-1]
for (i in 2:(n-1)) {
  p[i] <- q[i] - q[i-1]
}
chisq.test(Y, p = p)

File: Chap1.py
# Chap1 Python代码
# 自动从chap1.html同步生成

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.arange(0, 10)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# 概率分布列
for p, color, ls in [(0.1, 'C0', '-'), (0.2, 'C1', '--'),
                      (0.3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[0].plot(x, stats.binom.pmf(x, 9, p), color=color, ls=ls,
                 label=f'p={p}')
axes[0].set_xlabel('k'); axes[0].set_ylabel('p(k)')
axes[0].set_title('二项分布概率质量函数'); axes[0].legend()

# 累积分布函数
for p, color, ls in [(0.1, 'C0', '-'), (0.2, 'C1', '--'),
                      (0.3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[1].plot(x, stats.binom.cdf(x, 9, p), color=color, ls=ls,
                 label=f'p={p}')
axes[1].set_xlabel('k'); axes[1].set_ylabel('F(k)')
axes[1].set_title('二项分布累积分布函数'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.arange(0, 11)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

for lam, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                        (3, 'C2', '-.'), (5, 'C3', ':')]:
    axes[0].plot(x, stats.poisson.pmf(x, lam), color=color, ls=ls,
                 label=f'λ={lam}')
axes[0].set_xlabel('k'); axes[0].set_ylabel('p(k)')
axes[0].set_title('泊松分布概率质量函数'); axes[0].legend()

for lam, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                        (3, 'C2', '-.'), (5, 'C3', ':')]:
    axes[1].plot(x, stats.poisson.cdf(x, lam), color=color, ls=ls,
                 label=f'λ={lam}')
axes[1].set_xlabel('k'); axes[1].set_ylabel('F(k)')
axes[1].set_title('泊松分布累积分布函数'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.arange(0, 11)
fig, ax = plt.subplots(figsize=(8, 5))

for beta, color, ls in [(0.1, 'C0', '-'), (0.2, 'C1', '--'),
                         (0.3, 'C2', '-.'), (0.5, 'C3', ':')]:
    p = 1 / (1 + beta)
    ax.plot(x, stats.nbinom.pmf(x, 2, p), color=color, ls=ls,
            label=f'β={beta}')
ax.set_xlabel('k'); ax.set_ylabel('p(k)')
ax.set_title('负二项分布概率质量函数'); ax.legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.arange(0, 11)
p_nb = stats.nbinom.pmf(x, 4, 0.7)
p0 = p_nb[0]

# 零截断
pt = np.zeros_like(p_nb)
pt[1:] = p_nb[1:] / (1 - p0)

# 零调整
pm0 = 0.3
pm = np.zeros_like(p_nb)
pm[0] = pm0
pm[1:] = (1 - pm0) * p_nb[1:] / (1 - p0)

fig, ax = plt.subplots(figsize=(10, 5))
width = 0.35
ax.bar(x - width/2, p_nb, width, label='负二项', alpha=0.7)
ax.bar(x + width/2, pt, width, label='零截断负二项', alpha=0.7)
ax.set_xlabel('k'); ax.set_ylabel('概率')
ax.set_title('负二项分布与零截断负二项分布')
ax.set_xticks(x); ax.legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0, 5, 500)
fig, ax = plt.subplots(figsize=(8, 5))
for rate, color, ls in [(0.5, 'C0', '-'), (1, 'C1', '--'),
                         (2, 'C2', '-.'), (5, 'C3', ':')]:
    ax.plot(x, stats.expon.pdf(x, scale=1/rate), color=color, ls=ls,
            label=f'rate={rate}')
ax.set_xlabel('x'); ax.set_ylabel('f(x)')
ax.set_title('指数分布概率密度函数'); ax.legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0, 4, 1000)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

for shape, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                          (3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[0].plot(x, stats.gamma.pdf(x, shape, scale=1), color=color, ls=ls,
                 label=f'shape={shape}')
axes[0].set_xlabel('x'); axes[0].set_ylabel('f(x)')
axes[0].set_title('伽马分布（改变形状参数）'); axes[0].legend()

for scale, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                          (3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[1].plot(x, stats.gamma.pdf(x, 1, scale=scale), color=color, ls=ls,
                 label=f'scale={scale}')
axes[1].set_xlabel('x'); axes[1].set_ylabel('f(x)')
axes[1].set_title('伽马分布（改变尺度参数）'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0.01, 5, 500)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# 改变均值参数
for mu, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                       (5, 'C2', '-.'), (0.5, 'C3', ':')]:
    lam = 1
    y = np.sqrt(lam/(2*np.pi*x**3)) * np.exp(-lam*(x-mu)**2/(2*mu**2*x))
    axes[0].plot(x, y, color=color, ls=ls, label=f'mu={mu}')
axes[0].set_xlabel('x'); axes[0].set_ylabel('f(x)')
axes[0].set_title('逆高斯分布（改变均值）'); axes[0].legend()

# 改变形状参数
for lam, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                        (5, 'C2', '-.'), (0.5, 'C3', ':')]:
    mu = 1
    y = np.sqrt(lam/(2*np.pi*x**3)) * np.exp(-lam*(x-mu)**2/(2*mu**2*x))
    axes[1].plot(x, y, color=color, ls=ls, label=f'shape={lam}')
axes[1].set_xlabel('x'); axes[1].set_ylabel('f(x)')
axes[1].set_title('逆高斯分布（改变形状）'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt

def pareto_pdf(x, alpha, lam):
    return alpha * lam**alpha / (lam + x)**(alpha + 1)

x = np.linspace(0, 3, 300)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

for alpha, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                          (5, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[0].plot(x, pareto_pdf(x, alpha, 3), color=color, ls=ls,
                 label=f'α={alpha}')
axes[0].set_xlabel('x'); axes[0].set_ylabel('f(x)')
axes[0].set_title('帕累托分布（改变形状参数）'); axes[0].legend()

for lam, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                        (5, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[1].plot(x, pareto_pdf(x, 2, lam), color=color, ls=ls,
                 label=f'λ={lam}')
axes[1].set_xlabel('x'); axes[1].set_ylabel('f(x)')
axes[1].set_title('帕累托分布（改变尺度参数）'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0, 7, 500)
fig, ax = plt.subplots(figsize=(8, 5))
for sdlog, color, ls in [(0.5, 'C0', '-'), (1, 'C1', '--'),
                          (3, 'C2', '-.'), (10, 'C3', ':')]:
    ax.plot(x, stats.lognorm.pdf(x, s=sdlog, scale=np.exp(1)),
            color=color, ls=ls, label=f'sdlog={sdlog}')
ax.set_xlabel('x'); ax.set_ylabel('f(x)')
ax.set_title('对数正态分布'); ax.legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0, 3, 500)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

for shape, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                          (3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[0].plot(x, stats.weibull_min.pdf(x, shape, scale=1),
                 color=color, ls=ls, label=f'shape={shape}')
axes[0].set_xlabel('x'); axes[0].set_ylabel('f(x)')
axes[0].set_title('威布尔分布（改变形状参数）'); axes[0].legend()

for scale, color, ls in [(1, 'C0', '-'), (2, 'C1', '--'),
                          (3, 'C2', '-.'), (0.5, 'C3', ':')]:
    axes[1].plot(x, stats.weibull_min.pdf(x, 1, scale=scale),
                 color=color, ls=ls, label=f'scale={scale}')
axes[1].set_xlabel('x'); axes[1].set_ylabel('f(x)')
axes[1].set_title('威布尔分布（改变尺度参数）'); axes[1].legend()
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

x = np.linspace(0.001, 1, 1000)
f_mix = 0.3 * stats.lognorm.pdf(x, s=2, scale=np.exp(1)) + \
        0.7 * stats.lognorm.pdf(x, s=4, scale=np.exp(3))
f1 = stats.lognorm.pdf(x, s=2, scale=np.exp(1))
f2 = stats.lognorm.pdf(x, s=4, scale=np.exp(3))

fig, ax = plt.subplots(figsize=(8, 5))
ax.plot(x, f_mix, 'r-', lw=2, label='混合分布')
ax.plot(x, f1, 'b--', label='LN(1,2)')
ax.plot(x, f2, 'g-.', label='LN(3,4)')
ax.set_xlabel('x'); ax.set_ylabel('f(x)')
ax.set_title('混合分布'); ax.legend()
ax.set_ylim(0, 2)
plt.tight_layout(); plt.show()

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

np.random.seed(123)
x = np.random.gamma(shape=2, scale=1/2.5, size=100)

# MLE
shape_mle, loc_mle, scale_mle = stats.gamma.fit(x, floc=0)
print(f"伽马分布MLE: shape={shape_mle:.4f}, scale={scale_mle:.4f}, rate={1/scale_mle:.4f}")

# 矩估计
xbar = np.mean(x)
s2 = np.var(x, ddof=1)
shape_mme = xbar**2 / s2
scale_mme = s2 / xbar
print(f"伽马分布MME: shape={shape_mme:.4f}, scale={scale_mme:.4f}")

# 绘图
fig, ax = plt.subplots(figsize=(8, 5))
ax.hist(x, bins=20, density=True, alpha=0.5, label='数据')
xx = np.linspace(0, max(x), 200)
ax.plot(xx, stats.gamma.pdf(xx, shape_mle, scale=scale_mle),
        'r-', lw=2, label='MLE拟合')
ax.plot(xx, stats.gamma.pdf(xx, shape_mme, scale=scale_mme),
        'b--', lw=2, label='MME拟合')
ax.set_xlabel('x'); ax.set_ylabel('f(x)')
ax.set_title('伽马分布拟合'); ax.legend()
plt.tight_layout(); plt.show()

import numpy as np
from scipy import stats

# 卡方拟合优度检验：维修费用数据
lower = np.array([0, 1000, 2000, 3000, 4000, 5000])
upper = np.array([1000, 2000, 3000, 4000, 5000, np.inf])
freq = np.array([250, 300, 250, 150, 100, 0])

# 指数分布MLE
midpoints = np.array([500, 1500, 2500, 3500, 4500, 5500])
n_total = freq.sum()
xbar = np.sum(freq * midpoints) / n_total
lambda_hat = 1 / xbar

# 期望频数
p_exp = stats.expon.cdf(upper, scale=1/lambda_hat) - \
        stats.expon.cdf(lower, scale=1/lambda_hat)
exp_freq = n_total * p_exp

# 卡方统计量
chi_sq = np.sum((freq - exp_freq)**2 / exp_freq)
df = len(freq) - 1 - 1
p_value = 1 - stats.chi2.cdf(chi_sq, df)
print(f"卡方拟合优度检验（指数分布）:")
print(f"  lambda_hat = {lambda_hat:.6f}")
print(f"  chi_sq = {chi_sq:.4f}, df = {df}, p-value = {p_value:.4f}")

# 泊松分布检验
X = np.arange(0, 7)
Y = np.array([7, 10, 12, 8, 3, 2, 0])
lambda_hat_pois = np.sum(X * Y) / np.sum(Y)

p_pois = np.zeros(len(Y))
p_pois[0] = stats.poisson.cdf(0, lambda_hat_pois)
p_pois[-1] = 1 - stats.poisson.cdf(len(Y)-2, lambda_hat_pois)
for i in range(1, len(Y)-1):
    p_pois[i] = stats.poisson.cdf(X[i], lambda_hat_pois) - \
                stats.poisson.cdf(X[i]-1, lambda_hat_pois)

exp_freq_pois = np.sum(Y) * p_pois
chi_sq_pois = np.sum((Y - exp_freq_pois)**2 / exp_freq_pois)
df_pois = len(Y) - 1 - 1
p_value_pois = 1 - stats.chi2.cdf(chi_sq_pois, df_pois)
print(f"\n卡方拟合优度检验（泊松分布）:")
print(f"  lambda_hat = {lambda_hat_pois:.4f}")
print(f"  chi_sq = {chi_sq_pois:.4f}, df = {df_pois}, p-value = {p_value_pois:.4f}")